test/unit/app/services/asset_manager/attachment_updater_test.rb

```diff
@@ -4,281 +4,234 @@ class AssetManager::AttachmentUpdaterTest < ActiveSupport::TestCase
-        let(:draft) { false }
+      it "marks corresponding assets as not draft when the attachment has been replaced" do
+        attachment.attachment_data.replace_with!(AttachmentData.create!(file: File.open(fixture_path.join("sample.docx"))))
 
-        it "marks corresponding assets as not draft" do
-          update_service.expects(:call)
-                          .with(original_asset.asset_manager_id, attachment_data, nil, { "draft" => false })
-          update_service.expects(:call)
-                        .with(thumbnail_asset.asset_manager_id, attachment_data, nil, { "draft" => false })
+        expected_attribute_hash = {
+          "draft" => false,
+        }
 
-          updater.call(attachment_data, draft_status: true)
+        assets.each do |asset|
+          AssetManager::AssetUpdater.expects(:call).with(asset.asset_manager_id, attachment.attachment_data, nil, expected_attribute_hash)
         end
+
+        AssetManager::AttachmentUpdater.call(attachment.attachment_data, draft_status: true)
       end
+    end
 
-      context "and attachment is replaced" do
-        let(:replaced) { true }
+    context "when the attachment's attachable is a draft and is access limited" do
+      it "updates the access limited state of all assets" do
+        edition = create(:draft_news_article, :access_limited)
+        attachment = create(:file_attachment, attachable: edition)
+        assets = [
+          Asset.new(asset_manager_id: "original_asset_manager_id", variant: Asset.variants[:original], filename: "simple.pdf"),
+          Asset.new(asset_manager_id: "thumbnail_asset_manager_id", variant: Asset.variants[:thumbnail], filename: "thumbnail_simple.pdf.png"),
+        ]
+        attachment.attachment_data.assets = assets
 
-        it "marks corresponding assets as not draft even though attachment is draft" do
-          update_service.expects(:call)
-                          .with(original_asset.asset_manager_id, attachment_data, nil, { "draft" => false })
-          update_service.expects(:call)
-                        .with(thumbnail_asset.asset_manager_id, attachment_data, nil, { "draft" => false })
+        expected_attribute_hash = {
+          "access_limited_organisation_ids" => edition.organisations.map(&:content_id),
+        }
 
-          updater.call(attachment_data, draft_status: true)
+        assets.each do |asset|
+          AssetManager::AssetUpdater.expects(:call).with(asset.asset_manager_id, attachment.attachment_data, nil, expected_attribute_hash)
         end
+
+        AssetManager::AttachmentUpdater.call(attachment.attachment_data, access_limited: true)
       end
     end
 
-    describe "link header updates" do
-      let(:edition) { FactoryBot.create(:published_edition) }
-      let(:parent_document_url) { edition.public_url }
-      let(:attachment) { FactoryBot.create(:file_attachment, file: simple_pdf, attachable: edition) }
+    context "when attachment's attachable is published" do
```

Comment:
Most of this context now wraps the replacement logic tests but I don't think there is a strict correlation there.

Reply:
In fact, thinking about it, I would expect replacement only to apply for drafts? Presumably once the attachable has been published the user has to create a new edition before they can replace the attachment data?

Reply:
That makes more sense yes. I think most updates are in draft, except for the ones triggered by the edition listeners, which are then caused by some user action such as publishing etc.

Reply:
Having done a bit of further digging, I've found that the replacement logic is only invoked during bulk uploads. That user interface is only available when an edition is in draft state so I think testing with a draft attachable is appropriate.